### PR TITLE
Add about section

### LINKED ISSUE #12: Allow for custom colours in spell instructions
A new section could be used to specify metadata like `config:` or `about:` and the option to specify colour could be in there.

### PR DESCRIPTION
**Pull request by AndycptSher**
#12
implemented about section

### CODE REVIEW

File: docs/spell_attributes_list.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+TODO: format
+
```
Comment (CocytusDEDI) [inline docs/spell_attributes_list.txt]:
```suggestion
```

File: src/lib.rs [hunk 2]
```diff
@@ -636,10 +636,17 @@ impl Spell {
 
     fn internal_check_allowed_to_cast(instructions: Vec<u64>) -> Result<(), &'static str> {
         let mut instructions_iter = instructions.iter();
+        let mut section: Option<u64> = None;
         while let Some(&bits) = instructions_iter.next() {
+            if let Some(502) = section { // ignore all checks in metadata section
+                continue;
+            }
```
Comment (CocytusDEDI) [inline src/lib.rs]: Section can never be rewritten to after section becomes 502

File: src/lib.rs [hunk 3]
```diff
@@ -719,10 +729,38 @@ impl Spell {
...
             },
+            502 => {
+                self.set_meta_data(section_instructions.clone())
+            },
             _ => panic!("Invalid section")
         }
     }
 
+    fn set_meta_data(&mut self, equations: Vec<u64>) {
+        let mut codes = equations.into_iter();
```
Comment (CocytusDEDI) [inline src/lib.rs]: Equations should be renamed to attributes

```suggestion
    fn set_meta_data(&mut self, attributes: Vec<u64>) {
        let mut codes = attributes.into_iter();
```

File: src/lib.rs [hunk 4]
```diff
@@ -719,10 +729,38 @@ impl Spell {
...
+                    .map(|x| f64::from_bits(x) as f32)
+                    .collect::<Vec<f32>>()[..] {
+                        [c, d, e] => self.color = Color{r: c, g: d, b: e, a: SPELL_TRANSPARENCY},
+                        _ => panic!("Invalid data")
+                    }
+                },
+                1 => { // set destroy_when_done
+                    codes.next();
+                    panic!("Implementation not specified for destroy_when_done")
+                },
```
Comment (CocytusDEDI) [inline src/lib.rs]: destroy_when_done is planned yet
```suggestion
```

File: src/lib.rs [hunk 5]
```diff
@@ -719,10 +729,38 @@ impl Spell {
...
+            match code {
+                0 => { // set colour
+                    match match vec![codes.next(), codes.next(), codes.next()].into_iter().collect::<Option<Vec<u64>>>(){ // transpose vec of option into option of vec
+                        Some(n) => n,
+                        None => panic!("Invalid data")
+                    }.into_iter()
+                    .map(|x| f64::from_bits(x) as f32)
+                    .collect::<Vec<f32>>()[..] {
+                        [c, d, e] => self.color = Color{r: c, g: d, b: e, a: SPELL_TRANSPARENCY},
+                        _ => panic!("Invalid data")
```
Comment (CocytusDEDI) [inline src/lib.rs]:
```suggestion
                0 => { // Set colour
                    match match vec![codes.next(), codes.next(), codes.next()].into_iter().collect::<Option<Vec<u64>>>(){ // Transpose vec of option into option of vec
                        Some(color_vector) => color_vector,
                        None => panic!("Invalid data: There should be three color values")
                    }.into_iter()
                    .map(|x| f64::from_bits(x) as f32)
                    .collect::<Vec<f32>>()[..] {
                        [red, green, blue] => self.color = Color{r: red, g: green, b: blue, a: SPELL_TRANSPARENCY},
                        _ => panic!("Failed to parse colors")
```

File: src/lib.rs [hunk 6]
```diff
@@ -719,10 +729,38 @@ impl Spell {
...
+                    }
+                },
+                1 => { // set destroy_when_done
+                    codes.next();
+                    panic!("Implementation not specified for destroy_when_done")
+                },
+                _ => {}
+            }
+        }
+
```
Comment (CocytusDEDI) [inline src/lib.rs]:
```suggestion
```

File: src/spelltranslator.rs [hunk 7]
```diff
@@ -1,11 +1,13 @@
 use lazy_static::lazy_static;
 use std::collections::HashMap;
+use core::num::ParseFloatError;
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: Don't use ParseFloatError, as the rest of the program just uses `&'static str`

File: src/spelltranslator.rs [hunk 8]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
     }
 }
 
+fn parse_equation(equation: &str) -> Result<Vec<u64>, &'static str>{
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: Better name needed
```suggestion
fn parse_about_line(equation: &str) -> Result<Vec<u64>, &'static str>{
```

File: src/spelltranslator.rs [hunk 9]
```diff
@@ -80,6 +83,8 @@ pub fn parse_spell(spell_code: &str) -> Result<Vec<u64>, &'static str> {
                 expected_closing_brackets -= 1;
             } else if trimmed_line == "" {
                 continue
+            } else if trimmed_line.contains('=') && Some(502) == in_section { // Indicates an assignment of metadata
+                instructions.extend(parse_equation(trimmed_line)?)
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]:
```suggestion
                instructions.extend(parse_about_line(trimmed_line)?)
```

File: src/spelltranslator.rs [hunk 10]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
     }
 }
 
+fn parse_equation(equation: &str) -> Result<Vec<u64>, &'static str>{
+    let (name, value) = match equation.split_once('='){
+        Some(n) => n,
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: Name variables better, shouldn't be just `n`

File: src/spelltranslator.rs [hunk 11]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
...
+        None => return Err("equation not valid")
+    };
+    
+    match (name.trim(), value.trim()) { // because no map for tuples in rust
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]:
```suggestion
    match (name.trim(), value.trim()) {
```

File: src/spelltranslator.rs [hunk 12]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
...
+            let numbers = match match match match values.strip_prefix('[')
+            .and_then(|x| x.strip_suffix(']')) {
+                Some(x) => x,
+                None => {
+                    return Err("Invalid parameters: should be a list and have \"[\" \"]\"")
+                }
+            }.split(',')
+            .map(str::trim)
+            .map(str::parse::<f32>)
+            .collect::<Result<Vec<f32>, ParseFloatError>>() {
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: Don't use ParseFloatError

File: src/spelltranslator.rs [hunk 13]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
...
+                Ok(n) => n,
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: Name variables better than just `n`

File: src/spelltranslator.rs [hunk 14]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
...
+fn parse_equation(equation: &str) -> Result<Vec<u64>, &'static str>{
+    let (name, value) = match equation.split_once('='){
+        Some(n) => n,
+        None => return Err("equation not valid")
+    };
+    
+    match (name.trim(), value.trim()) { // because no map for tuples in rust
+        ("colour", values) | ("color", values) => {
+            // returns used to sidestep borrowing rules
+            let numbers = match match match match values.strip_prefix('[')
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: Having four match statements is insane behaviour, and makes it very unreadable. This needs to be refactored.

File: src/spelltranslator.rs [hunk 15]
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
...
+            }.into_iter()
+            .map(|x| f64::to_bits(x as f64))
+            .collect::<Vec<u64>>();
+            Ok(vec![0].into_iter().chain(numbers.into_iter()).collect())
+        },
+        ("destroy_when_done", boolean) => match boolean {
+            "true" => Ok(vec![1, 1]),
+            "false" => Ok(vec![1, 0]),
+            _ => Err("Invalid parameter")
+        },
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]: destroy_when_done isn't planned yet
```suggestion
```

File: src/spelltranslator.rs [hunk 16]
```diff
@@ -694,6 +746,33 @@ mod tests {
...
+    }
+    
+    #[test]
+    fn parse_colour_attribution() {
+        assert_eq!(parse_equation("colour = [0.4, 0, 0.8]"), Ok(vec![0, f64::to_bits((0.4 as f32) as f64), 0, f64::to_bits((0.8 as f32) as f64)]));
+        assert_eq!(parse_equation("color = [0.212, 1, 2.3]"), Err("Invalid values: arguments should be between 0 and 1"));
+    }
+
+    #[test]
+    fn parse_limit_colour_attribution() {
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]:
```suggestion
    fn parse_limit_colour() {
```

File: src/spelltranslator.rs [hunk 17]
```diff
@@ -694,6 +746,33 @@ mod tests {
...
+        assert_eq!(parse_equation("     color      =        [   0.212,    1,0.3]"), Ok(vec![0, f64::to_bits((0.212 as f32) as f64), f64::to_bits((1 as f32) as f64), f64::to_bits((0.3 as f32) as f64)]));
+    }
+
+    #[test]
+    fn parse_bool_attrtbution() {
+        assert_eq!(parse_equation("destroy_when_done = true"), Ok(vec![1, 1]))
+    }
```
Comment (CocytusDEDI) [inline src/spelltranslator.rs]:
```suggestion
```

**Review by CocytusDEDI — CHANGES_REQUESTED** [review]
Overall major refactors need to be done for readability and one bug fix. After it has been refactored testing will be done.

**Comment by AndycptSher** [comment]
## Note: possible bug
current implementation permits if statements in about section introducing possible parsing errors as additional opcodes have not been accounted for

**Comment by CocytusDEDI** [comment]
Can confirm that's a bug, I'll be fixing it along with a couple other things.

**Pull request merged**